crates/matrix-sdk-crypto/src/olm/group_sessions/share_strategy.rs

```diff
@@ -0,0 +1,508 @@
+        ///     - It is signed by it's owner identity, and this identity has
+        ///       been trusted via interactive verification.
+        ///     - It is the current own device of the user.
+        only_allow_trusted_devices: bool,
+    }, // XXX some new strategy to be defined later
+}
+
+impl CollectStrategy {
+    /// Creates a new legacy strategy, based on per device trust.
+    pub const fn new_device_based(only_allow_trusted_devices: bool) -> Self {
+        CollectStrategy::DeviceBasedStrategy { only_allow_trusted_devices }
+    }
+}
+
+impl Default for CollectStrategy {
+    fn default() -> Self {
+        CollectStrategy::new_device_based(false)
+    }
+}
+
+impl CollectRecipientsHelper {
+    pub(crate) async fn collect_session_recipients(
+        settings: &EncryptionSettings,
+        store: &Store,
+        users: impl Iterator<Item = &UserId>,
+        outbound: &OutboundGroupSession,
+    ) -> OlmResult<CollectRecipientsResult> {
+        let users: BTreeSet<&UserId> = users.collect();
+
+        let users_shared_with: BTreeSet<OwnedUserId> =
+            outbound.shared_with_set.read().unwrap().keys().cloned().collect();
+
+        let users_shared_with: BTreeSet<&UserId> =
+            users_shared_with.iter().map(Deref::deref).collect();
+
+        // A user left if a user is missing from the set of users that should
+        // get the session but is in the set of users that received the session.
+        let user_left = !users_shared_with.difference(&users).collect::<BTreeSet<_>>().is_empty();
+
+        let visibility_changed =
+            outbound.settings().history_visibility != settings.history_visibility;
+        let algorithm_changed = outbound.settings().algorithm != settings.algorithm;
+
+        let sharing_strategy_changed =
+            outbound.settings().sharing_strategy != settings.sharing_strategy;
+
+        let mut should_rotate =
+            user_left || visibility_changed || algorithm_changed || sharing_strategy_changed;
+
+        if should_rotate {
+            debug!(
+                should_rotate,
+                user_left,
+                visibility_changed,
+                algorithm_changed,
+                "Rotating room key to protect room history",
+            );
+        }
+
+        let sharing_result = match settings.sharing_strategy {
+            CollectStrategy::DeviceBasedStrategy { only_allow_trusted_devices } => {
+                Self::collect_session_recipients_device_based(
+                    store,
+                    users,
+                    only_allow_trusted_devices,
+                )
+                .await?
+            }
+        };
+
+        // If we haven't already concluded that the session should be
+        // rotated for other reasons, we also need to check whether any
+        // of the devices in the session got deleted or blacklisted in the
+        // meantime. If so, we should also rotate the session.
+        if !should_rotate {
```

Comment:
I would prefer if the contents of this `if` were a separate function - the containing function is very long.